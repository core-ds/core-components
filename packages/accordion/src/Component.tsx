import React, {
    AnchorHTMLAttributes,
    FC,
    KeyboardEvent,
    ReactNode,
    useCallback,
    useState,
} from 'react';
import cn from 'classnames';

import { Typography } from '@alfalab/core-components-typography';

import { DefaultControlIcon } from './components';
import { useRecalculateContentHeight } from './hooks';
import { ControlPosition } from './typings';

import styles from './index.module.css';

export type AccordionProps = {
    /**
     * Состояние компонента
     */
    expanded?: boolean;

    /**
     * Элемент заголовка
     */
    header: ReactNode;

    /**
     * Слот для элемента управления
     */
    control?: ReactNode;

    /**
     * Указывает компоненту - где будет размещен control
     */
    controlPosition?: ControlPosition;

    /**
     * Начальное состояние uncontrolled компонента
     */
    defaultExpanded?: boolean;

    /**
     * Основной элемент для отображения содержимого
     */
    children?: ReactNode;

    /**
     * Дополнительный класс обертки
     */
    className?: string;

    /**
     * Дополнительный класс для контейнера с заголовком
     */
    containerClassName?: string;

    /**
     * Дополнительный класс для header
     */
    headerClassName?: string;

    /**
     * Дополнительный класс для control
     */
    controlClassName?: string;

    /**
     * Дополнительный класс для body
     */
    bodyClassName?: string;

    /**
     * Обработчик смены состояний `expanded`
     */
    onExpandedChange?: (expanded: boolean) => void;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    dataTestId?: string;
} & AnchorHTMLAttributes<HTMLDivElement>;

export const Accordion: FC<AccordionProps> = ({
    expanded,
    defaultExpanded = false,
    header,
    control,
    controlPosition = 'end',
    children,
    className,
    containerClassName,
    headerClassName,
    controlClassName,
    bodyClassName,
    onExpandedChange,
    dataTestId,
    ...rest
}) => {
    const uncontrolled = expanded === undefined;
    const [expandedState, setExpanded] = useState(uncontrolled ? defaultExpanded : expanded);
    const isExpanded = uncontrolled ? expandedState : expanded;

    const isStartPosition = controlPosition === 'start';

    const { contentRef, contentCaseRef } = useRecalculateContentHeight(isExpanded);

    const controlContent =
        control === undefined ? (
            <DefaultControlIcon expanded={isExpanded} startPosition={isStartPosition} />
        ) : (
            control
        );

    const headerContent =
        typeof header === 'string' ? (
            <Typography.Text view='primary-large' weight='medium'>
                {header}
            </Typography.Text>
        ) : (
            header
        );

    const bodyContent =
        typeof children === 'string' ? (
            <Typography.Text view='primary-medium'>{children}</Typography.Text>
        ) : (
            children
        );

    const handleExpandedChange = useCallback(() => {
        if (uncontrolled) {
            setExpanded(!isExpanded);
        }

        onExpandedChange?.(!isExpanded);
    }, [isExpanded, onExpandedChange, uncontrolled]);

    const handleKeyDown = (event: KeyboardEvent) => {
        if (event.key === 'Enter') {
            handleExpandedChange();
        }
    };

    return (
        <div {...rest} data-test-id={dataTestId} className={cn(styles.accordion, className)}>
            <div
                role='button'
                tabIndex={0}
                aria-expanded={isExpanded}
                onClick={handleExpandedChange}
                onKeyDown={handleKeyDown}
                className={cn(styles.container, containerClassName)}
            >
                <div
                    className={cn(styles.header, headerClassName, {
                        [styles.endPosition]: isStartPosition,
                    })}
                >
                    {headerContent}
                </div>

                <div
                    className={cn(styles.control, controlClassName, {
                        [styles.startPosition]: isStartPosition,
                    })}
                >
                    {controlContent}
                </div>
            </div>

            <div
                ref={contentRef}
                className={cn(styles.body, bodyClassName, { [styles.expandedBody]: isExpanded })}
            >
                <div className={cn(styles.bodyContent)} ref={contentCaseRef}>
                    {bodyContent}
                </div>
            </div>
        </div>
    );
};
