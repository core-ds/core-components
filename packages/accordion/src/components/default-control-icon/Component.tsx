import React, { FC } from 'react';
import cn from 'classnames';

import { ChevronDownMIcon } from '@alfalab/icons-glyph/ChevronDownMIcon';

import styles from './index.module.css';

type Props = {
    expanded: boolean;
    startPosition: boolean;
};
export const DefaultControlIcon: FC<Props> = ({ expanded, startPosition }) => (
    <ChevronDownMIcon
        className={cn(styles.icon, {
            [styles.expanded]: expanded,
            [styles.startPosition]: startPosition,
        })}
    />
);
