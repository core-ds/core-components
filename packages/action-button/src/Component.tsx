import React, {
    AnchorHTMLAttributes,
    ButtonHTMLAttributes,
    forwardRef,
    Fragment,
    useRef,
} from 'react';
import mergeRefs from 'react-merge-refs';
import cn from 'classnames';

import { Loader } from '@alfalab/core-components-loader';
import { useFocus } from '@alfalab/hooks';

import { useLoader } from './hooks';

import defaultColors from './default.module.css';
import styles from './index.module.css';
import invertedColors from './inverted.module.css';
import staticColors from './static.module.css';

const colorStyles = {
    default: defaultColors,
    inverted: invertedColors,
    static: staticColors,
};

/**
 * Минимальное время отображения лоадера - 500мс,
 * чтобы при быстрых ответах от сервера кнопка не «моргала».
 */
const LOADER_MIN_DISPLAY_INTERVAL = 500;

type Colors = 'default' | 'inverted' | 'static';

type ComponentProps = {
    /**
     * Иконка кнопки
     */
    icon: React.ReactNode;

    /**
     *  Размер кнопки
     */
    size?: 's';

    /**
     * Тип кнопки
     */
    view?: 'primary' | 'secondary';

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Дополнительный класс для обертки иконки
     */
    iconWrapperClassName?: string;

    /**
     * Значение href для ссылки
     */
    href?: string;

    /**
     * Заблокировать кнопку
     */
    disabled?: boolean;

    /**
     * Показать лоадер
     */
    loading?: boolean;

    /**
     * Id компонента для тестов
     */
    dataTestId?: string;

    /**
     * Палитра, в контексте которой используется кнопка
     */
    colors?: Colors;
};

type AnchorProps = ComponentProps & AnchorHTMLAttributes<HTMLAnchorElement>;
type ButtonProps = ComponentProps & ButtonHTMLAttributes<HTMLButtonElement>;
export type ActionButtonProps = Partial<AnchorProps | ButtonProps>;

export const ActionButton = forwardRef<HTMLAnchorElement | HTMLButtonElement, ActionButtonProps>(
    (
        {
            className,
            icon,
            children,
            href,
            size = 's',
            view = 'primary',
            type = 'button',
            iconWrapperClassName,
            disabled,
            loading,
            dataTestId,
            colors = 'default',
            ...rest
        },
        ref,
    ) => {
        const componentRef = useRef<HTMLElement>(null);

        const [focused] = useFocus(componentRef, 'keyboard');

        const { showLoader } = useLoader(!!loading, LOADER_MIN_DISPLAY_INTERVAL);

        const componentProps = {
            className: cn(
                styles.component,
                colorStyles[colors][view],
                styles[size],
                {
                    [styles.focused]: focused,
                    [styles.disabled]: disabled,
                    [styles.loading]: showLoader,
                },
                className,
            ),
            'data-test-id': dataTestId,
        };

        const buttonChildren = (
            <Fragment>
                <span
                    role='img'
                    className={cn(
                        styles.iconWrapper,
                        colorStyles[colors].iconWrapper,
                        styles[size],
                        iconWrapperClassName,
                    )}
                >
                    {showLoader ? <Loader dataTestId='loader' /> : icon}
                </span>
                <span className={styles.label}>{children}</span>
            </Fragment>
        );

        if (href) {
            return (
                <a
                    role='button'
                    ref={mergeRefs([componentRef, ref])}
                    href={href}
                    aria-disabled={disabled || loading}
                    {...componentProps}
                    {...(rest as AnchorHTMLAttributes<HTMLAnchorElement>)}
                >
                    {buttonChildren}
                </a>
            );
        }

        return (
            <button
                ref={mergeRefs([componentRef, ref])}
                // eslint-disable-next-line react/button-has-type
                type={type as ButtonHTMLAttributes<HTMLButtonElement>['type']}
                disabled={disabled || loading}
                {...componentProps}
                {...(rest as ButtonHTMLAttributes<HTMLButtonElement>)}
            >
                {buttonChildren}
            </button>
        );
    },
);
