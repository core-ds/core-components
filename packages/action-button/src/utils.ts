import { getDataTestId } from '@alfalab/core-components-shared';

export function getActionButtonTestIds(dataTestId: string) {
    return {
        actionButton: dataTestId,
        spinner: getDataTestId(dataTestId, 'loader'),
    };
}
