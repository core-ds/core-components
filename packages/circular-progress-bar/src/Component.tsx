import React, { ElementType, ReactNode, useMemo } from 'react';
import cn from 'classnames';

import { Typography } from '@alfalab/core-components-typography';

import styles from './index.module.css';

const SIZES = {
    xs: 24,
    s: 48,
    m: 64,
    l: 80,
    xl: 128,
    xxl: 144,
};

const STROKE = {
    xs: 4,
    s: 4,
    m: 6,
    l: 8,
    xl: 10,
    xxl: 12,
};

const VIEW_TITLE = {
    xs: 'small',
    s: 'small',
    m: 'small',
    l: 'xsmall',
    xl: 'medium',
    xxl: 'medium',
} as const;

const VIEW_TEXT = {
    xs: 'secondary-small',
    s: 'secondary-small',
    m: 'secondary-large',
    l: 'secondary-large',
    xl: 'secondary-large',
    xxl: 'secondary-large',
} as const;

export type CircularProgressBarProps = {
    /**
     * Уровень прогресса, %
     */
    value: number;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Основной текст
     */
    title?: ReactNode;

    /**
     * Цвет контента
     */
    contentColor?: 'primary' | 'secondary' | 'tertiary' | 'positive' | 'negative';

    /**
     * Дополнительный текст
     */
    subtitle?: ReactNode;

    /**
     * Основной текст при 100%
     */
    titleComplete?: ReactNode;

    /**
     * Дополнительный текст при 100%
     */
    subtitleComplete?: ReactNode;

    /**
     * Цвет заполнения
     */
    view?: 'positive' | 'negative';

    /**
     * Размер (xxl — 144×144px, xl — 128×128px, l — 80×80px, m — 64×64px, s — 48×48px, xs — 24×24px)
     */
    size?: 'xxl' | 'xl' | 'l' | 'm' | 's' | 'xs';

    /**
     * Наличие желоба
     */
    stroke?: boolean;

    /**
     * Заливка при 100%
     */
    fillComplete?: boolean;

    /**
     * Цвет текста при 100%
     */
    completeTextColor?: 'primary' | 'primary-inverted' | 'positive' | 'negative';

    /**
     * Цвет иконки при 100%
     */
    completeIconColor?: 'primary-inverted' | 'positive' | 'negative' | 'tertiary';

    /**
     * Компонент иконки
     */
    icon?: ElementType<{ className?: string }>;

    /**
     * Компонент иконки при 100%
     */
    iconComplete?: ElementType<{ className?: string }>;

    /**
     * Направление прогресса (clockwise - по часовой стрелке, counter-clockwise - против часовой стрелки)
     */
    direction?: 'clockwise' | 'counter-clockwise';

    /**
     * Высота компонента, min = 24; max = 144
     * использовать совместно с size :
     * xxl от 144
     * xl  от 128 до 143
     * l   от 80 до 127
     * m   от 64 до 79
     * s   от 48 до 63
     * xs  от 24 до 47
     */
    height?: number;

    /**
     * Id компонента для тестов
     */
    dataTestId?: string;

    /**
     * Дочерние элементы.
     */
    children?: ReactNode;
};

/**
 * Компонент круглого прогресс бара.
 */
export const CircularProgressBar: React.FC<CircularProgressBarProps> = ({
    value,
    view = 'positive',
    size = 'm',
    className,
    dataTestId,
    title = value ? value.toString() : '0',
    titleComplete,
    subtitle,
    contentColor = 'secondary',
    subtitleComplete,
    stroke = true,
    fillComplete,
    icon: Icon,
    iconComplete: IconComplete,
    completeTextColor,
    completeIconColor = 'tertiary',
    direction = 'clockwise',
    height,
    children,
}) => {
    const memorized = useMemo(() => {
        const strokeWidth = STROKE[size];
        const maxProgress = 100;
        const minProgress = 0;
        const widthSVG = SIZES[size];
        const heightSVG = SIZES[size];
        const center = widthSVG / 2;
        const radius = center - strokeWidth / 2;
        const circumference = Math.PI * radius * 2;
        const progress = Math.min(Math.max(value, minProgress), maxProgress);
        const strokeDasharray = circumference.toFixed(3);
        const strokeDashoffset = (((100 - progress) / 100) * circumference).toFixed(3);

        return {
            widthSVG,
            heightSVG,
            center,
            radius,
            strokeDasharray,
            strokeDashoffset,
        };
    }, [value, size]);

    const isComplete = value === 100;
    const isCompleteTextColor = isComplete && completeTextColor;
    const titleContent = titleComplete && isComplete ? titleComplete : title;
    const subtitleContent = subtitleComplete && isComplete ? subtitleComplete : subtitle;
    const IconComponent = IconComplete && isComplete ? IconComplete : Icon;

    const renderTitleString = () =>
        SIZES[size] > 64 ? (
            <Typography.TitleMobile
                className={cn(styles.typography, styles.title)}
                color={isCompleteTextColor ? completeTextColor : contentColor}
                tag='div'
                font='system'
                view={VIEW_TITLE[size]}
            >
                {titleContent}
            </Typography.TitleMobile>
        ) : (
            <Typography.Text
                className={styles.title}
                color={isCompleteTextColor ? completeTextColor : contentColor}
                tag='div'
                weight='bold'
                view={VIEW_TEXT[size]}
            >
                {titleContent}
            </Typography.Text>
        );

    const renderTitle = () => (typeof title === 'string' ? renderTitleString() : titleContent);

    const renderSubTitle = () =>
        typeof subtitle === 'string' ? (
            <Typography.Text
                tag='div'
                className={styles.subtitle}
                color={isCompleteTextColor ? completeTextColor : contentColor}
                view='primary-small'
            >
                {subtitleContent}
            </Typography.Text>
        ) : (
            subtitleContent
        );

    const renderIcon = () => (
        <span
            className={cn(
                styles.iconWrapper,
                styles[size],
                styles.tertiary,
                styles[`icon-${contentColor}`],
                {
                    [styles[`icon-${completeIconColor}`]]: completeIconColor,
                },
            )}
        >
            {IconComponent && <IconComponent className={styles.icon} />}
        </span>
    );

    const renderContent = () =>
        Icon || (IconComplete && isComplete) ? (
            renderIcon()
        ) : (
            <React.Fragment>
                {SIZES[size] > 24 && renderTitle()}
                {SIZES[size] > 64 && renderSubTitle()}
            </React.Fragment>
        );

    return (
        <div
            className={cn(styles.component, styles[size], className)}
            style={{
                ...(height && { height, width: height }),
            }}
            data-test-id={dataTestId}
        >
            <svg
                viewBox={`0 0 ${memorized.widthSVG} ${memorized.heightSVG}`}
                className={styles.svg}
                xmlns='http://www.w3.org/2000/svg'
            >
                <circle
                    className={cn(styles.backgroundCircle, styles[size], {
                        [styles.stroke]: !stroke,
                    })}
                    cx={memorized.center}
                    cy={memorized.center}
                    r={memorized.radius}
                    strokeWidth={STROKE[size]}
                />
                <circle
                    className={cn(styles.progressCircle, styles[view], styles[size], {
                        [styles[`bg-${view}`]]: fillComplete && isComplete,
                    })}
                    cx={memorized.center}
                    cy={memorized.center}
                    r={memorized.radius}
                    strokeWidth={STROKE[size]}
                    strokeDasharray={memorized.strokeDasharray}
                    strokeDashoffset={
                        direction === 'counter-clockwise'
                            ? -memorized.strokeDashoffset
                            : memorized.strokeDashoffset
                    }
                    transform={`rotate(${-90} ${memorized.center} ${memorized.center})`}
                />
            </svg>
            <div
                className={cn(styles.labelWrapper, {
                    [styles.label]: Icon || IconComplete,
                })}
            >
                {children || renderContent()}
            </div>
        </div>
    );
};
