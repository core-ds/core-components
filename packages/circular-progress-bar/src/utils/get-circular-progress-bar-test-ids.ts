import { getDataTestId } from '@alfalab/core-components-shared';

export function getCircularProgressBarTestIds(dataTestId: string) {
    return {
        progressBar: dataTestId,
        title: getDataTestId(dataTestId, 'title'),
        subtitle: getDataTestId(dataTestId, 'subtitle'),
        circleProgressBar: getDataTestId(dataTestId, 'circle-progress-bar'),
        circleProgressValue: getDataTestId(dataTestId, 'circle-progress-value'),
    };
}
