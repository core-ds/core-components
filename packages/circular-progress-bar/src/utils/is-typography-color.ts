import { TYPOGRAPHY_COLOR } from '../consts';
import { type TypographyColor } from '../types/typography-color';

export const isTypographyColor = (color: TypographyColor | string): color is TypographyColor =>
    TYPOGRAPHY_COLOR.includes(color);
