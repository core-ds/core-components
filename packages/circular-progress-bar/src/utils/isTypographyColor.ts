import { TYPOGRAPHY_COLOR } from '../consts';
import { TypographyColor } from '../types/typography-color';

export const isTypographyColor = (color: TypographyColor | string): color is TypographyColor =>
    TYPOGRAPHY_COLOR.includes(color);
