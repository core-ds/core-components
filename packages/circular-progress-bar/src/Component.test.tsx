import React from 'react';
import { cleanup, render, screen } from '@testing-library/react';

import { CircularProgressBar } from './Component';
import { getCircularProgressBarTestIds } from './utils/get-circular-progress-bar-test-ids';
import { act } from 'react-dom/test-utils';

describe('ProgressBar', () => {
    describe('Snapshots tests', () => {
        it('should match snapshot', () => {
            const { container } = render(<CircularProgressBar value={20} />);
            expect(container).toMatchSnapshot();
        });

        it('should match with view snapshot', () => {
            const { container } = render(<CircularProgressBar value={20} view='negative' />);
            expect(container).toMatchSnapshot();
        });

        it('should match custom title & subtitle snapshot', () => {
            const { container } = render(
                <CircularProgressBar
                    value={20}
                    view='negative'
                    title={<span style={{ color: 'red' }}>title</span>}
                    subtitle={<span style={{ color: 'green' }}>subtitle</span>}
                />,
            );
            expect(container).toMatchSnapshot();
        });
    });

    describe('Attributes tests', () => {
        it('should set `data-test-id` attribute', () => {
            const testId = 'test-id';
            const { getByTestId, queryByTestId } = render(
                <CircularProgressBar value={20} dataTestId={testId} subtitle='subtitle' />,
            );

            const testIds = getCircularProgressBarTestIds(testId);

            expect(getByTestId(testId)).toBeInTheDocument();
            expect(getByTestId(testIds.progressBar)).toBeInTheDocument();
            expect(getByTestId(testIds.title)).toBeInTheDocument();
            expect(queryByTestId(testIds.subtitle)).not.toBeInTheDocument();
            expect(getByTestId(testIds.circleProgressBar)).toBeInTheDocument();
            expect(getByTestId(testIds.circleProgressValue)).toBeInTheDocument();
        });

        it('should set `data-test-id` attribute for size more than 64', () => {
            const testId = 'test-id';
            const { getByTestId } = render(
                <CircularProgressBar
                    value={20}
                    size={80}
                    dataTestId={testId}
                    subtitle='subtitle'
                />,
            );

            const testIds = getCircularProgressBarTestIds(testId);

            expect(getByTestId(testIds.title)).toBeInTheDocument();
            expect(getByTestId(testIds.subtitle)).toBeInTheDocument();
        });

        it('should set `data-test-id` attribute for timer', () => {
            const testId = 'test-id';
            const { getByTestId } = render(
                <CircularProgressBar timer={true} value={20} dataTestId={testId} />,
            );

            const testIds = getCircularProgressBarTestIds(testId);

            expect(getByTestId(testIds.title)).toBeInTheDocument();
        });

        it('should set `data-test-id` attribute for timer with size more than 64', () => {
            const testId = 'test-id';
            const { getByTestId } = render(
                <CircularProgressBar timer={true} value={20} size={80} dataTestId={testId} />,
            );

            const testIds = getCircularProgressBarTestIds(testId);

            expect(getByTestId(testIds.title)).toBeInTheDocument();
        });

        it('should use passed `value`', () => {
            const { container } = render(<CircularProgressBar value={72} size={80} />);

            expect(container.querySelector('.progressCircle')).toHaveAttribute(
                'stroke-dashoffset',
                '63.335',
            );
        });

        it('should use passed `title`', () => {
            const text = 'test-text';
            const { container } = render(<CircularProgressBar value={72} title={text} />);

            expect(container.querySelector('.title')).toHaveTextContent(text);
        });

        it('should `title = value` if `title = undefined`', () => {
            const value = 72;
            const { container } = render(<CircularProgressBar value={value} />);

            expect(container.querySelector('.title')).toHaveTextContent(`${value}`);
        });

        it('should use passed `subtitle`', () => {
            const text = 'test-text';
            const { container } = render(
                <CircularProgressBar value={72} size={80} subtitle={text} />,
            );

            expect(container.querySelector('.subtitle')).toHaveTextContent(text);
        });

        it('should use only passed `children`', () => {
            const subtitle = 'test-subtitle';
            const text = 'test-text';
            const { container } = render(
                <CircularProgressBar value={72} subtitle={subtitle}>
                    {text}
                </CircularProgressBar>,
            );

            expect(container.querySelector('.labelWrapper')).toHaveTextContent(text);
            expect(container.querySelector('.subtitle')).not.toBeInTheDocument();
        });
    });

    describe('Classes tests', () => {
        it('should use default `size`', () => {
            const { container } = render(<CircularProgressBar value={20} />);

            expect(container.firstElementChild).toHaveClass('size-64');
        });

        it('should use passed `size`', () => {
            const { container } = render(<CircularProgressBar value={20} size={80} />);

            expect(container.firstElementChild).toHaveClass('size-80');
        });

        it('should set `className` class to root', () => {
            const className = 'test-class';
            const { container } = render(<CircularProgressBar value={20} className={className} />);

            expect(container.firstElementChild).toHaveClass(className);
        });

        it('should set default color class', () => {
            const { container } = render(<CircularProgressBar value={20} />);

            expect(container.querySelector('.progressCircle')).toHaveClass('positive');
        });

        it('should set color class', () => {
            const { container } = render(<CircularProgressBar value={20} view='negative' />);

            expect(container.querySelector('.progressCircle')).toHaveClass('negative');
        });
    });

    it('should unmount without errors', () => {
        const { unmount } = render(<CircularProgressBar value={20} />);

        expect(unmount).not.toThrowError();
    });

    describe('Timer tests', () => {
        jest.useFakeTimers();
        afterEach(cleanup);

        it('should set "timer" props', () => {
            render(<CircularProgressBar timer={true} value={3599} />);

            const text = screen.queryByText('59:59');

            expect(text).toBeInTheDocument();
        });

        it('should set max value', () => {
            render(<CircularProgressBar timer={true} value={999999} />);

            const text = screen.queryByText('59:59');

            expect(text).toBeInTheDocument();
        });

        it('should set min value', () => {
            render(<CircularProgressBar timer={true} value={-999999} />);

            const text = screen.queryByText('0:00');

            expect(text).toBeInTheDocument();
        });

        it('should "backward" counting', () => {
            render(<CircularProgressBar timer={true} value={60} />);

            expect(screen.queryByText('1:00')).toBeInTheDocument();

            act(() => {
                jest.advanceTimersByTime(15000);
            });

            expect(screen.queryByText('0:45')).toBeInTheDocument();

            act(() => {
                jest.advanceTimersByTime(45000);
            });

            expect(screen.queryByText('0:00')).toBeInTheDocument();

            act(() => {
                jest.advanceTimersByTime(10000);
            });

            expect(screen.queryByText('0:00')).toBeInTheDocument();
        });

        it('should "forward" counting', () => {
            render(<CircularProgressBar timer={true} value={60} counting='forward' />);

            expect(screen.queryByText('0:00')).toBeInTheDocument();

            act(() => {
                jest.advanceTimersByTime(15000);
            });

            expect(screen.queryByText('0:15')).toBeInTheDocument();

            act(() => {
                jest.advanceTimersByTime(45000);
            });

            expect(screen.queryByText('1:00')).toBeInTheDocument();

            act(() => {
                jest.advanceTimersByTime(10000);
            });

            expect(screen.queryByText('1:00')).toBeInTheDocument();
        });

        it('should "desc" directionType', () => {
            const testId = 'test-id';

            render(<CircularProgressBar dataTestId={testId} timer={true} value={60} size={80} />);

            const testIds = getCircularProgressBarTestIds(testId);

            const circle = screen.getByTestId(testIds.circleProgressValue);

            expect(circle.getAttribute('stroke-dasharray')).toEqual('226.195');
            expect(circle.getAttribute('stroke-dashoffset')).toEqual('0.000'); // круг полностью заполнен

            act(() => {
                jest.advanceTimersByTime(15000);
            });

            expect(circle.getAttribute('stroke-dasharray')).toEqual('226.195');
            expect(circle.getAttribute('stroke-dashoffset')).toEqual('56.549'); // круг заполнен на 3/4

            act(() => {
                jest.advanceTimersByTime(45000);
            });

            expect(circle.getAttribute('stroke-dasharray')).toEqual('226.195');
            expect(circle.getAttribute('stroke-dashoffset')).toEqual('226.195'); // круг пустой
        });

        it('should "asc" directionType', () => {
            const testId = 'test-id';

            render(
                <CircularProgressBar
                    dataTestId={testId}
                    timer={true}
                    value={60}
                    size={80}
                    directionType='asc'
                />,
            );

            const testIds = getCircularProgressBarTestIds(testId);

            const circle = screen.getByTestId(testIds.circleProgressValue);

            expect(circle.getAttribute('stroke-dasharray')).toEqual('226.195');
            expect(circle.getAttribute('stroke-dashoffset')).toEqual('226.195'); // круг пустой

            act(() => {
                jest.advanceTimersByTime(15000);
            });

            expect(circle.getAttribute('stroke-dasharray')).toEqual('226.195');
            expect(circle.getAttribute('stroke-dashoffset')).toEqual('169.646'); // круг заполнен на 1/4

            act(() => {
                jest.advanceTimersByTime(45000);
            });

            expect(circle.getAttribute('stroke-dasharray')).toEqual('226.195');
            expect(circle.getAttribute('stroke-dashoffset')).toEqual('0.000'); // круг полностью заполнен
        });
    });

    describe('Color tests', () => {
        it('should set contentColor classname', () => {
            const testId = 'test-id';
            render(
                <CircularProgressBar
                    value={100}
                    size={80}
                    dataTestId={testId}
                    title='Title'
                    subtitle='Subtitle'
                    contentColor='primary'
                />,
            );

            const testIds = getCircularProgressBarTestIds(testId);
            const title = screen.getByTestId(testIds.title);
            const subtitle = screen.getByTestId(testIds.subtitle);

            expect(title).toHaveClass('primary');
            expect(subtitle).toHaveClass('primary');
        });

        it('should override contentColor classname to titleColor classname', () => {
            const testId = 'test-id';
            render(
                <CircularProgressBar
                    value={100}
                    size={80}
                    dataTestId={testId}
                    title='Title'
                    subtitle='Subtitle'
                    contentColor='primary'
                    titleColor='positive'
                />,
            );

            const testIds = getCircularProgressBarTestIds(testId);
            const title = screen.getByTestId(testIds.title);
            const subtitle = screen.getByTestId(testIds.subtitle);

            expect(title).toHaveClass('positive');
            expect(subtitle).toHaveClass('primary');
        });

        it('should override contentColor classname to subtitleColor classname', () => {
            const testId = 'test-id';
            render(
                <CircularProgressBar
                    value={100}
                    size={80}
                    dataTestId={testId}
                    title='Title'
                    subtitle='Subtitle'
                    contentColor='primary'
                    subtitleColor='positive'
                />,
            );

            const testIds = getCircularProgressBarTestIds(testId);
            const title = screen.getByTestId(testIds.title);
            const subtitle = screen.getByTestId(testIds.subtitle);

            expect(title).toHaveClass('primary');
            expect(subtitle).toHaveClass('positive');
        });
    });

    describe('Inline color tests', () => {
        it('should set "contentColor"', () => {
            const testId = 'test-id';

            render(
                <CircularProgressBar
                    value={100}
                    size={80}
                    dataTestId={testId}
                    title='Title'
                    subtitle='Subtitle'
                    contentColor='tomato'
                />,
            );

            const testIds = getCircularProgressBarTestIds(testId);
            const title = screen.getByTestId(testIds.title);
            const subtitle = screen.getByTestId(testIds.subtitle);

            expect(title).toHaveStyle('color: tomato');
            expect(subtitle).toHaveStyle('color: tomato');
        });

        it('should override "contentColor" with "titleColor"', () => {
            const testId = 'test-id';

            render(
                <CircularProgressBar
                    value={100}
                    size={80}
                    dataTestId={testId}
                    title='Title'
                    subtitle='Subtitle'
                    contentColor='tomato'
                    titleColor='blue'
                />,
            );

            const testIds = getCircularProgressBarTestIds(testId);
            const title = screen.getByTestId(testIds.title);
            const subtitle = screen.getByTestId(testIds.subtitle);

            expect(title).toHaveStyle('color: blue');
            expect(subtitle).toHaveStyle('color: tomato');
        });

        it('should override "contentColor" with "subtitleColor"', () => {
            const testId = 'test-id';

            render(
                <CircularProgressBar
                    value={100}
                    size={80}
                    dataTestId={testId}
                    title='Title'
                    subtitle='Subtitle'
                    contentColor='tomato'
                    subtitleColor='blue'
                />,
            );

            const testIds = getCircularProgressBarTestIds(testId);
            const subtitle = screen.getByTestId(testIds.subtitle);
            const title = screen.getByTestId(testIds.title);

            expect(subtitle).toHaveStyle('color: blue');
            expect(title).toHaveStyle('color: tomato');
        });
    });
});
