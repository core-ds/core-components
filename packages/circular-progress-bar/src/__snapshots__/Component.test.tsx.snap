// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar Snapshots tests should match custom title & subtitle snapshot 1`] = `
<div>
  <div
    class="component m"
  >
    <svg
      class="svg"
      viewBox="0 0 64 64"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        class="backgroundCircle m"
        cx="32"
        cy="32"
        r="29"
        stroke-width="6"
      />
      <circle
        class="progressCircle negative m"
        cx="32"
        cy="32"
        r="29"
        stroke-dasharray="182.212"
        stroke-dashoffset="145.770"
        stroke-width="6"
        transform="rotate(-90 32 32)"
      />
    </svg>
    <div
      class="labelWrapper"
    >
      <span
        style="color: red;"
      >
        title
      </span>
    </div>
  </div>
</div>
`;

exports[`ProgressBar Snapshots tests should match snapshot 1`] = `
<div>
  <div
    class="component m"
  >
    <svg
      class="svg"
      viewBox="0 0 64 64"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        class="backgroundCircle m"
        cx="32"
        cy="32"
        r="29"
        stroke-width="6"
      />
      <circle
        class="progressCircle positive m"
        cx="32"
        cy="32"
        r="29"
        stroke-dasharray="182.212"
        stroke-dashoffset="145.770"
        stroke-width="6"
        transform="rotate(-90 32 32)"
      />
    </svg>
    <div
      class="labelWrapper"
    >
      <div
        class="title secondary secondary-large bold"
      >
        20
      </div>
    </div>
  </div>
</div>
`;

exports[`ProgressBar Snapshots tests should match with view snapshot 1`] = `
<div>
  <div
    class="component m"
  >
    <svg
      class="svg"
      viewBox="0 0 64 64"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        class="backgroundCircle m"
        cx="32"
        cy="32"
        r="29"
        stroke-width="6"
      />
      <circle
        class="progressCircle negative m"
        cx="32"
        cy="32"
        r="29"
        stroke-dasharray="182.212"
        stroke-dashoffset="145.770"
        stroke-width="6"
        transform="rotate(-90 32 32)"
      />
    </svg>
    <div
      class="labelWrapper"
    >
      <div
        class="title secondary secondary-large bold"
      >
        20
      </div>
    </div>
  </div>
</div>
`;
